File: src/engine/editor/board.cc
#include "board.hh"

#include <iostream>

#include "engine/sandbox/sandbox.hh"

size_t Board::board_counter_ = 0;

Board::Board(intpos_t w, intpos_t h, ComponentDatabase const& component_db, SandboxRecompilationFn recompile)
    : id_(board_counter_++), w_(w), h_(h), component_db_(component_db), recompile_(recompile)
{
}

Board::Board(json const& content, ComponentDatabase const& component_db, SandboxRecompilationFn recompile)
    : id_(content.at("id")), w_(content.at("w")), h_(content.at("h")), component_db_(component_db), recompile_(recompile)
{
    if (board_counter_ <= id_)
        board_counter_ = id_ + 1;

    for (auto const& jwire: content.at("wires").items())
        wires_.emplace(Position(jwire.key()), Wire(jwire.value()));

    for (auto const& jcomp: content.at("components").items()) {
        Position pos(jcomp.key());
        Component* component = add_component(jcomp.value().at("name"), pos.x, pos.y, true);
        if (jcomp.value().contains("value"))
            component->def->unserialize_component(*component, jcomp.value().at("value"));
    }

    recompile_();
}

Component* Board::add_component(std::string const& component_name, intpos_t x, intpos_t y, bool bypass_reset)
{
    // TODO - check - is there a component here already?
    auto it = components_.emplace(Position { this, x, y, Direction::Center }, component_db_.create_component(component_name));
    if (!bypass_reset)
        recompile_();
    return &it.first->second;
}

void Board::draw_wire(Wire::Width width, Wire::Layer layer, intpos_t x0, intpos_t y0, intpos_t x1, intpos_t y1, Orientation orientation)
{
    wire_management_.start_drawing({ this, x0, y0 }, width, layer);
    for (auto const& [pos, wire] : wire_management_.stop_drawing({ this, x1, y1 }))
        wires_.emplace(pos, wire);
    recompile_();
}

void Board::merge_wires(std::map<Position, Wire> const& wires)
{
    for (auto const& [pos, wire]: wires)
        wires_.emplace(pos, wire);
    recompile_();
}

void Board::clear_tile(intpos_t x, intpos_t y)
{
    wires_.erase({ this, x, y, Direction::Center });
    for (Direction dir: DIRECTIONS)
        wires_.erase({ this, x, y, dir });
    components_.erase({ this, x, y });
    recompile_();
}

void Board::clear()
{
    wires_.clear();
    components_.clear();
    recompile_();
}

void Board::rotate_component(intpos_t x, intpos_t y)
{
    auto it = components_.find({ id_, x, y });
    if (it != components_.end()) {
        Component& component = it->second;
        if (component.def->type == ComponentDefinition::Type::SingleTile) {
            if (component.def->can_rotate)
                component.rotation = dir_rotate_component(component.rotation);
        } else {
            throw std::runtime_error("IC rotation not supported yet.");  // TODO - implement for ICs
        }
    }
    recompile_();
}

json Board::serialize() const
{
    json jwires = json::object();
    for (auto const& [pos, wire] : wires_)
        jwires[std::to_string(pos)] = std::to_string(wire);

    json jcomponents = json::object();
    for (auto const& [pos, component]: components_)
        jcomponents[std::to_string(pos)] = component.def->serialize(component);

    return {
        { "id", id_ },
        { "w", w_ },
        { "h", h_ },
        { "wires", jwires },
        { "components", jcomponents },
    };
}

File: engine/board.lua
Board = {}
Board.__index = Board

function Board.new(w, h, sandbox)
   local self = setmetatable({}, Board)
   self.w = w
   self.h = h
   self.wires = {}
   self.components = {}
   self.sandbox = sandbox
   return self
end

function Board:wire(pos)
   return self.wires[pos:hash()]
end

function Board:add_wire(pos, wire)
   if pos.x < 0 or pos.y < 0 or pos.x >= self.w or pos.y >= self.h then return end
   self.wires[pos:hash()] = wire
   self:remove_wires_under_ics()
end

function Board:add_wires(pos1, pos2, orientation, wire)
   for _,pos in ipairs(Position.a_to_b(pos1, pos2, orientation)) do
      if pos.x >= 0 or pos.y >= 0 or pos.x < self.w or pos.y < self.h then
         self.wires[pos:hash()] = wire
      end
   end
   self:remove_wires_under_ics()
end

function Board:add_component(key, pos, dir)
   assert(dir ~= CENTER)
   
   local def = self.sandbox.component_db:def(key)
   assert(def ~= nil, "Component '" .. key .. "' not found in database.")
   
   -- TODO is it inside the board
   
   -- TODO is there another component there
   
   -- initialize component
   local component = Component.new(def, pos, dir)
   self.components[#self.components+1] = component
   
   if component.def.type ~= "single_tile" then
      self:remove_wires_for_ic(component:rect())
   end
end

function Board:clear_tile(pos)
   assert(pos.dir == CENTER)
   
   -- TODO clear component
   
   -- clear wires
   self.wires[P(pos.x, pos.y, N):hash()] = nil
   self.wires[P(pos.x, pos.y, E):hash()] = nil
   self.wires[P(pos.x, pos.y, S):hash()] = nil
   self.wires[P(pos.x, pos.y, W):hash()] = nil
end

function Board:remove_wires_under_ics()
   for _,c in ipairs(self.components) do
      if c.def.type == "single_tile" then
         self:remove_wires_for_ic(c:rect())
      end
   end
end

function Board:remove_wires_for_ic(rect)
   -- TODO
end